src/kudu/consensus/replicate_msg_wrapper.h

```diff
@@ -0,0 +1,190 @@
+
+  ReplicateRefPtr GetCompressedMsg() const { return compressed_msg_; }
+
+private:
+
+  Status UncompressMsg(faststring *buffer = nullptr) {
+    DCHECK(!msg_ && compressed_msg_);
+
+    const OperationType op_type = compressed_msg_->get()->op_type();
+    const WritePayloadPB &payload = compressed_msg_->get()->write_payload();
+    const CompressionType compression_codec = payload.compression_codec();
+    const int64_t uncompressed_size = payload.uncompressed_size();
+    const int64_t compressed_size = payload.payload().size();
+
+    if (!buffer) {
+      buffer = new faststring(uncompressed_size);
+    }
+
+    VLOG(2) << "Uncompressing message"
+            << " opid: " << compressed_msg_->get()->id().ShortDebugString()
+            << " codec: " << compression_codec << " op_type: " << op_type
+            << " compressed payload size: " << compressed_size
+            << " uncompressed payload size: " << uncompressed_size;
+
+    // Resize buffer to hold uncompressed payload.
+    // TODO: needs perf testing and maybe implement streaming (un)compression
+    buffer->resize(uncompressed_size);
+    Slice compressed_slice(payload.payload().c_str(), compressed_size);
+
+    Status status;
+    if (codec_ && codec_->type() == compression_codec) {
+      status = codec_->Uncompress(compressed_slice, buffer->data(),
+                                  uncompressed_size);
+    } else {
+      // Either compression is not enabled on this instance OR this message uses
+      // a different compression codec. Get the right codec now
+      status = GetCompressionCodec(compression_codec, &codec_);
+      if (status.ok()) {
+        status = codec_->Uncompress(compressed_slice, buffer->data(),
+                                    uncompressed_size);
+      }
+    }
+
+    // Return early if uncompression failed
+    RETURN_NOT_OK_PREPEND(
+        status,
+        strings::Substitute(
+            "Failed to uncompress OpId $0. Compression codec used: $1, "
+            "Operation type: $2, Compressed payload size: $3"
+            "Uncompressed payload size: $4",
+            compressed_msg_->get()->id().ShortDebugString(), compression_codec,
+            op_type, compressed_size, uncompressed_size));
+
+    // Now create a new ReplicateMsg and copy over the contents from the
+    // original msg and the uncompressed payload
+    std::unique_ptr<ReplicateMsg> rep_msg(new ReplicateMsg);
+    *(rep_msg->mutable_id()) = compressed_msg_->get()->id();
+    rep_msg->set_timestamp(compressed_msg_->get()->timestamp());
+    rep_msg->set_op_type(compressed_msg_->get()->op_type());
+
+    WritePayloadPB *write_payload = rep_msg->mutable_write_payload();
+    write_payload->set_payload(buffer->ToString());
+
+    msg_ = make_scoped_refptr_replicate(rep_msg.release());
+    return Status::OK();
+  }
+
+  Status CompressMsg(faststring *buffer = nullptr) {
+    DCHECK(msg_ && !compressed_msg_);
+
+    if (!should_compress_) {
+      return Status::OK();
```

Comment:
this creates a problem IMO. Since some messages like Rotate events/control messages are not compressed, their compressed_msg field will not  be populated. How is append entries RPC going to figure out what field to use to write to file

Reply:
If compressed msg != nullptr, it is the one that is added to the cache, otherwise the uncompressed msg is added to the cache. This logic is in LogCache::AppendEntries()